pkg/mcp/sink/server_sink.go

```diff
@@ -0,0 +1,126 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sink
+
+import (
+	"errors"
+	"io"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	mcp "istio.io/api/mcp/v1alpha1"
+	"istio.io/istio/pkg/mcp/internal"
+	"istio.io/istio/pkg/mcp/monitoring"
+	"istio.io/istio/pkg/mcp/source"
+)
+
+var (
+	// For the purposes of logging rate limiting authz failures, this controls how
+	// many authz failus are logs as a burst every AUTHZ_FAILURE_LOG_FREQ.
+	authzFailureLogBurstSize = internal.EnvInt("AUTHZ_FAILURE_LOG_BURST_SIZE", 1)
+
+	// For the purposes of logging rate limiting authz failures, this controls how
+	// frequently bursts of authz failures are logged.
+	authzFailureLogFreq = internal.EnvDur("AUTHZ_FAILURE_LOG_FREQ", time.Minute)
+)
+
+// AuthChecker is used to check the transport auth info that is associated with each stream. If the function
+// returns nil, then the connection will be allowed. If the function returns an error, then it will be
+// percolated up to the gRPC stack.
+//
+// Note that it is possible that this method can be called with nil authInfo. This can happen either if there
+// is no peer info, or if the underlying gRPC stream is insecure. The implementations should be resilient in
+// this case and apply appropriate policy.
+type AuthChecker interface {
+	Check(authInfo credentials.AuthInfo) error
+}
+
+// ServerSource implements the server implementation of the MCP sink service.
+type ServerSink struct {
+	authCheck                   AuthChecker
+	checkFailureRecordLimiter   *rate.Limiter
+	failureCountSinceLastRecord int
+	connections                 int64
+	reporter                    monitoring.Reporter
+	sink                        *Sink
+}
+
+var _ mcp.ResourceSinkServer = &ServerSink{}
+
+// NewServerSource creates a new instance of a MCP sink server.
+func NewServerSink(
+	supportedTypeURLs []string,
+	updater Updater,
+	id string,
+	metadata map[string]string,
+	reporter monitoring.Reporter,
+	authChecker source.AuthChecker) *ServerSink {
+	s := &ServerSink{
+		sink:      NewSink(supportedTypeURLs, updater, id, metadata, reporter),
+		authCheck: authChecker,
+		reporter:  reporter,
+	}
+
+	// Initialize record limiter for the auth checker.
+	limit := rate.Every(authzFailureLogFreq)
+	limiter := rate.NewLimiter(limit, authzFailureLogBurstSize)
+	s.checkFailureRecordLimiter = limiter
+
+	return s
+}
+
+// EstablishResourceStream implements the ResourceSinkServer interface.
+func (s *ServerSink) EstablishResourceStream(stream mcp.ResourceSink_EstablishResourceStreamServer) error {
+	// TODO support receiving configuration from multiple sources?
+	// TODO MVP - limit to one connection at a time?
+	if !atomic.CompareAndSwapInt64(&s.connections, 0, 1) {
```

Comment:
If ServerSink a thread-safe abstraction? I wasn't expecting it to be...

Reply:
It's minimally thread-safe. The majority of state is per-connection/stream. The extra atomic check here prevents the case of multiple sources connecting, which isn't supported (yet). Eventually,  Galley may status from multiple Pilot instances, in which case a single sink _server_ may have multiple source connections.